Flaky test: Test_TerraformRecipe_AzureStorage (text file busy)

## Bug information

This is from a failed scheduled functional test run.

```
=== Failed
=== FAIL: test/functional/shared/resources Test_TerraformRecipe_AzureStorage/deploy_testdata/corerp-resources-terraform-azurestorage.bicep (33.39s)
    rptest.go:272: running step 0 of 1: deploy testdata/corerp-resources-terraform-azurestorage.bicep
    deployexecutor.go:80: deploying deploy testdata/corerp-resources-terraform-azurestorage.bicep from file testdata/corerp-resources-terraform-azurestorage.bicep
    cli.go:468: [heartbeat] command rad deploy /home/runner/work/radius/radius/test/functional/shared/resources/testdata/corerp-resources-terraform-azurestorage.bicep --parameters moduleServer=http://tf-module-server.radius-test-tf-module-server.svc.cluster.local --parameters appName=corerp-resources-terraform-azstorage-app is still running after 10.000181348s
2023/10/24 08:49:30 Start streaming Kubernetes logs - Pod ctnr-live-ready-6d57b5d58c-ngxv4 is in state: Running
2023/10/24 08:49:31 Start streaming Kubernetes logs - Pod friendly-ctnr-567649468c-7r89p is in state: Running
2023/10/24 08:49:31 Streaming Kubernetes logs to /home/runner/work/radius/radius/dist/container_logs/friendly-ctnr-567649468c-7r89p.friendly-ctnr.log
2023/10/24 08:49:31 Start streaming Kubernetes logs - Pod friendly-ctnr-567649468c-7r89p is in state: Running
2023/10/24 08:49:31 Start streaming Kubernetes logs - Pod friendly-ctnr-6d9654d64c-smnvf is in state: Running
    cli.go:468: [heartbeat] command rad deploy /home/runner/work/radius/radius/test/functional/shared/resources/testdata/corerp-resources-terraform-azurestorage.bicep --parameters moduleServer=http://tf-module-server.radius-test-tf-module-server.svc.cluster.local --parameters appName=corerp-resources-terraform-azstorage-app is still running after 20.00[106](https://github.com/radius-project/radius/actions/runs/6624331998/job/17993310006#step:23:107)1603s
2023/10/24 08:49:41 Start streaming Kubernetes logs - Pod controller-95b8d7747-8hvvx is in state: Running
2023/10/24 08:49:41 Start streaming Kubernetes logs - Pod controller-95b8d7747-8hvvx is in state: Running
    cli.go:468: [heartbeat] command rad deploy /home/runner/work/radius/radius/test/functional/shared/resources/testdata/corerp-resources-terraform-azurestorage.bicep --parameters moduleServer=http://tf-module-server.radius-test-tf-module-server.svc.cluster.local/ --parameters appName=corerp-resources-terraform-azstorage-app is still running after 30.001184916s
2023/10/24 08:49:50 Saved container logs to /home/runner/work/radius/radius/dist/container_logs/friendly-ctnr-567649468c-7r89p.friendly-ctnr.log
2023/10/24 08:49:51 Start streaming Kubernetes logs - Pod friendly-ctnr-567649468c-7r89p is in state: Failed
2023/10/24 08:49:51 Start streaming Kubernetes logs - Pod friendly-ctnr-567649468c-7r89p is in state: Failed
2023/10/24 08:49:51 Start streaming Kubernetes logs - Pod friendly-ctnr-567649468c-7r89p is in state: Failed
2023/10/24 08:49:51 Start streaming Kubernetes logs - Pod friendly-ctnr-567649468c-7r89p is in state: Failed
2023/10/24 08:49:51 Start streaming Kubernetes logs - Pod friendly-rds-ctnr-58c85c867b-brkd7 is in state: Running
    cli.go:415: [rad] Building /home/runner/work/radius/radius/test/functional/shared/resources/testdata/corerp-resources-terraform-azurestorage.bicep...
    cli.go:415: [rad] Deploying template '/home/runner/work/radius/radius/test/functional/shared/resources/testdata/corerp-resources-terraform-azurestorage.bicep' into environment 'kind-radius' from workspace 'kind-radius'...
    cli.go:415: [rad] 
    cli.go:415: [rad] Deployment In Progress...
    cli.go:415: [rad] 
    cli.go:415: [rad] Error: {
    cli.go:415: [rad]   "code": "DeploymentFailed",
    cli.go:415: [rad]   "message": "At least one resource deployment operation failed. Please see the details for the specific operation that failed.",
    cli.go:415: [rad]   "target": "/planes/radius/local/resourceGroups/kind-radius/providers/Microsoft.Resources/deployments/rad-deploy-fbf2cc85-a4b9-401b-8ac6-df6ead7ae332",
    cli.go:415: [rad]   "details": [
    cli.go:415: [rad]     {
    cli.go:415: [rad]       "code": "OK",
    cli.go:415: [rad]       "message": "",
    cli.go:415: [rad]       "target": "/planes/radius/local/resourceGroups/kind-radius/providers/Applications.Core/environments/corerp-resources-terraform-azstorage-env"
    cli.go:415: [rad]     },
    cli.go:415: [rad]     {
    cli.go:415: [rad]       "code": "ResourceDeploymentFailure",
    cli.go:415: [rad]       "message": "Failed",
    cli.go:415: [rad]       "target": "/planes/radius/local/resourceGroups/kind-radius/providers/Applications.Core/extenders/corerp-resources-terraform-azstorage",
    cli.go:415: [rad]       "details": [
    cli.go:415: [rad]         {
    cli.go:415: [rad]           "code": "RecipeDeploymentFailed",
    cli.go:415: [rad]           "message": "code RecipeDownloadFailed: err failed to run terraform get to download the module from source \"http://tf-module-server.radius-test-tf-module-server.svc.cluster.local//azure-storage.zip\": fork/exec /terraform/9ffafbf8-5fa0-4286-9393-02ef37220196-18dc84ac-cd34-40e3-ab7e-53d356b60393/install/terraform: text file busy",
    cli.go:415: [rad]           "details": [
    cli.go:415: [rad]             {
    cli.go:415: [rad]               "code": "RecipeDownloadFailed",
    cli.go:415: [rad]               "message": "failed to run terraform get to download the module from source \"http://tf-module-server.radius-test-tf-module-server.svc.cluster.local//azure-storage.zip\": fork/exec /terraform/9ffafbf8-5fa0-4286-9393-02ef37220196-18dc84ac-cd34-40e3-ab7e-53d356b60393/install/terraform: text file busy"
    cli.go:415: [rad]             }
    cli.go:415: [rad]           ]
    cli.go:415: [rad]         }
    cli.go:415: [rad]       ]
    cli.go:415: [rad]     },
    cli.go:415: [rad]     {
    cli.go:415: [rad]       "code": "OK",
    cli.go:415: [rad]       "message": "",
    cli.go:415: [rad]       "target": "/planes/radius/local/resourceGroups/kind-radius/providers/Applications.Core/applications/corerp-resources-terraform-azstorage-app"
    cli.go:415: [rad]     }
    cli.go:415: [rad]   ]
    cli.go:415: [rad] }
    cli.go:415: [rad] 
    cli.go:415: [rad] TraceId:  8773b8c95cc25eaef5b0bb576315fd41
    cli.go:415: [rad] 
    cli.go:415: [rad] 
    deployexecutor.go:83: 
        	Error Trace:	/home/runner/work/radius/radius/test/step/deployexecutor.go:83
        	            				/home/runner/work/radius/radius/test/functional/shared/rptest.go:273
        	Error:      	Received unexpected error:
        	            	code DeploymentFailed: err At least one resource deployment operation failed. Please see the details for the specific operation that failed.
        	Test:       	Test_TerraformRecipe_AzureStorage/deploy_testdata/corerp-resources-terraform-azurestorage.bicep
        	Messages:   	failed to deploy deploy testdata/corerp-resources-terraform-azurestorage.bicep
    --- FAIL: Test_TerraformRecipe_AzureStorage/deploy_testdata/corerp-resources-terraform-azurestorage.bicep (33.39s)
```


[AB#9906](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9906)

--- a/pkg/metrics/recipeenginemetrics.go
+++ b/pkg/metrics/recipeenginemetrics.go
@@ -43,6 +43,9 @@ const (
 	// terraformInitializationDuration is the metric name for the Terraform initialization duration.
 	terraformInitializationDuration = "recipe.tf.init.duration"
 
+	// terraformVerficiationDuration is the metric name for the Terraform installation verification duration.
+	terraformInstallVerficiationDuration = "recipe.tf.install.verification.duration"
+
 	// RecipeEngineOperationExecute represents the Execute operation of the Recipe Engine.
 	RecipeEngineOperationExecute = "execute"
 
@@ -118,6 +121,14 @@ func (m *recipeEngineMetrics) RecordTerraformInstallationDuration(ctx context.Co
 	if m.valueRecorders[terraformInstallationDuration] != nil {
 		elapsedTime := float64(time.Since(startTime)) / float64(time.Millisecond)
 		m.valueRecorders[terraformInstallationDuration].Record(ctx, elapsedTime, metric.WithAttributes(attrs...))
+	}
+}
+
+// RecordTerraformInstallVerificationDuration records the duration of the Terraform installation verification with the given attributes.
+func (m *recipeEngineMetrics) RecordTerraformInstallVerificationDuration(ctx context.Context, startTime time.Time, attrs []attribute.KeyValue) {
+	if m.valueRecorders[terraformInstallVerficiationDuration] != nil {
+		elapsedTime := float64(time.Since(startTime)) / float64(time.Millisecond)
+		m.valueRecorders[terraformInstallVerficiationDuration].Record(ctx, elapsedTime, metric.WithAttributes(attrs...))
 	}
 }
 

--- a/pkg/metrics/recipeenginemetrics.go
+++ b/pkg/metrics/recipeenginemetrics.go
@@ -43,8 +43,8 @@ const (
 	// terraformInitializationDuration is the metric name for the Terraform initialization duration.
 	terraformInitializationDuration = "recipe.tf.init.duration"
 
-	// terraformInstallVerficiationDuration is the metric name for the Terraform installation verification duration.
-	terraformInstallVerficiationDuration = "recipe.tf.install.verification.duration"
+	// terraformInstallVerificationDuration is the metric name for verifying the completion of a Terraform installation duration.
+	terraformInstallVerificationDuration = "recipe.tf.install.verification.duration"
 
 	// RecipeEngineOperationExecute represents the Execute operation of the Recipe Engine.
 	RecipeEngineOperationExecute = "execute"
@@ -126,9 +126,9 @@ func (m *recipeEngineMetrics) RecordTerraformInstallationDuration(ctx context.Co
 
 // RecordTerraformInstallVerificationDuration records the duration of the Terraform installation verification with the given attributes.
 func (m *recipeEngineMetrics) RecordTerraformInstallVerificationDuration(ctx context.Context, startTime time.Time, attrs []attribute.KeyValue) {
-	if m.valueRecorders[terraformInstallVerficiationDuration] != nil {
+	if m.valueRecorders[terraformInstallVerificationDuration] != nil {
 		elapsedTime := float64(time.Since(startTime)) / float64(time.Millisecond)
-		m.valueRecorders[terraformInstallVerficiationDuration].Record(ctx, elapsedTime, metric.WithAttributes(attrs...))
+		m.valueRecorders[terraformInstallVerificationDuration].Record(ctx, elapsedTime, metric.WithAttributes(attrs...))
 	}
 }
 

--- a/pkg/metrics/recipeenginemetrics.go
+++ b/pkg/metrics/recipeenginemetrics.go
@@ -43,7 +43,7 @@ const (
 	// terraformInitializationDuration is the metric name for the Terraform initialization duration.
 	terraformInitializationDuration = "recipe.tf.init.duration"
 
-	// terraformVerficiationDuration is the metric name for the Terraform installation verification duration.
+	// terraformInstallVerficiationDuration is the metric name for the Terraform installation verification duration.
 	terraformInstallVerficiationDuration = "recipe.tf.install.verification.duration"
 
 	// RecipeEngineOperationExecute represents the Execute operation of the Recipe Engine.

--- a/pkg/recipes/terraform/execute.go
+++ b/pkg/recipes/terraform/execute.go
@@ -24,6 +24,7 @@ import (
 
 	install "github.com/hashicorp/hc-install"
 	"github.com/hashicorp/terraform-exec/tfexec"
+	"github.com/hashicorp/terraform-exec/tfexec"
 	tfjson "github.com/hashicorp/terraform-json"
 	"github.com/radius-project/radius/pkg/metrics"
 	"github.com/radius-project/radius/pkg/recipes"
@@ -70,7 +71,7 @@ func (e *executor) Deploy(ctx context.Context, options Options) (*tfjson.State, 
 
 	// Install Terraform
 	i := install.NewInstaller()
-	execPath, err := Install(ctx, i, options.RootDir)
+	tf, err := Install(ctx, i, options.RootDir)
 	// The terraform zip for installation is downloaded in a location outside of the install directory and is only accessible through the installer.Remove function -
 	// stored in latestVersion.pathsToRemove. So this needs to be called for complete cleanup even if the root terraform directory is deleted.
 	defer func() {
@@ -78,12 +79,6 @@ func (e *executor) Deploy(ctx context.Context, options Options) (*tfjson.State, 
 			logger.Info(fmt.Sprintf("Failed to cleanup Terraform installation: %s", err.Error()))
 		}
 	}()
-	if err != nil {
-		return nil, err
-	}
-
-	// Create a new instance of tfexec.Terraform with current Terraform installation path
-	tf, err := NewTerraform(ctx, options.RootDir, execPath)
 	if err != nil {
 		return nil, err
 	}
@@ -119,7 +114,7 @@ func (e *executor) Delete(ctx context.Context, options Options) error {
 
 	// Install Terraform
 	i := install.NewInstaller()
-	execPath, err := Install(ctx, i, options.RootDir)
+	tf, err := Install(ctx, i, options.RootDir)
 	// The terraform zip for installation is downloaded in a location outside of the install directory and is only accessible through the installer.Remove function -
 	// stored in latestVersion.pathsToRemove. So this needs to be called for complete cleanup even if the root terraform directory is deleted.
 	defer func() {
@@ -127,12 +122,6 @@ func (e *executor) Delete(ctx context.Context, options Options) error {
 			logger.Info(fmt.Sprintf("Failed to cleanup Terraform installation: %s", err.Error()))
 		}
 	}()
-	if err != nil {
-		return err
-	}
-
-	// Create a new instance of tfexec.Terraform with current Terraform installation path
-	tf, err := NewTerraform(ctx, options.RootDir, execPath)
 	if err != nil {
 		return err
 	}
@@ -179,7 +168,7 @@ func (e *executor) GetRecipeMetadata(ctx context.Context, options Options) (map[
 
 	// Install Terraform
 	i := install.NewInstaller()
-	execPath, err := Install(ctx, i, options.RootDir)
+	tf, err := Install(ctx, i, options.RootDir)
 	// The terraform zip for installation is downloaded in a location outside of the install directory and is only accessible through the installer.Remove function -
 	// stored in latestVersion.pathsToRemove. So this needs to be called for complete cleanup even if the root terraform directory is deleted.
 	defer func() {
@@ -187,12 +176,6 @@ func (e *executor) GetRecipeMetadata(ctx context.Context, options Options) (map[
 			logger.Info(fmt.Sprintf("Failed to cleanup Terraform installation: %s", err.Error()))
 		}
 	}()
-	if err != nil {
-		return nil, err
-	}
-
-	// Create a new instance of tfexec.Terraform with current Terraform installation path
-	tf, err := NewTerraform(ctx, options.RootDir, execPath)
 	if err != nil {
 		return nil, err
 	}
@@ -213,6 +196,7 @@ func (e *executor) GetRecipeMetadata(ctx context.Context, options Options) (map[
 }
 
 // generateConfig generates Terraform configuration with required inputs for the module, providers and backend to be initialized and applied.
+func (e *executor) generateConfig(ctx context.Context, tf *tfexec.Terraform, options Options) (string, error) {
 func (e *executor) generateConfig(ctx context.Context, tf *tfexec.Terraform, options Options) (string, error) {
 	logger := ucplog.FromContextOrDiscard(ctx)
 	workingDir := tf.WorkingDir()

--- a/pkg/recipes/terraform/execute.go
+++ b/pkg/recipes/terraform/execute.go
@@ -24,7 +24,6 @@ import (
 
 	install "github.com/hashicorp/hc-install"
 	"github.com/hashicorp/terraform-exec/tfexec"
-	"github.com/hashicorp/terraform-exec/tfexec"
 	tfjson "github.com/hashicorp/terraform-json"
 	"github.com/radius-project/radius/pkg/metrics"
 	"github.com/radius-project/radius/pkg/recipes"
@@ -196,7 +195,6 @@ func (e *executor) GetRecipeMetadata(ctx context.Context, options Options) (map[
 }
 
 // generateConfig generates Terraform configuration with required inputs for the module, providers and backend to be initialized and applied.
-func (e *executor) generateConfig(ctx context.Context, tf *tfexec.Terraform, options Options) (string, error) {
 func (e *executor) generateConfig(ctx context.Context, tf *tfexec.Terraform, options Options) (string, error) {
 	logger := ucplog.FromContextOrDiscard(ctx)
 	workingDir := tf.WorkingDir()

--- a/pkg/recipes/terraform/install.go
+++ b/pkg/recipes/terraform/install.go
@@ -94,6 +94,12 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (*
 		if err != nil {
 			if attempt < installRetryCount {
 				logger.Info(fmt.Sprintf(errMsg, err.Error(), retryDelaySecs))
+				metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallVerificationDuration(ctx, installStartTime,
+					[]attribute.KeyValue{
+						metrics.TerraformVersionAttrKey.String("latest"),
+						metrics.OperationStateAttrKey.String(metrics.FailedOperationState),
+					},
+				)
 				time.Sleep(time.Duration(retryDelaySecs) * time.Second)
 				continue
 			}
@@ -101,6 +107,14 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (*
 		}
 	}
 
+	metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallVerificationDuration(ctx, installStartTime,
+		[]attribute.KeyValue{
+			metrics.TerraformVersionAttrKey.String("latest"),
+			metrics.OperationStateAttrKey.String(metrics.SuccessfulOperationState),
+		},
+	)
+
+	// Configure Terraform logs once Terraform installation is complete
 	configureTerraformLogs(ctx, tf)
 
 	return tf, nil

--- a/pkg/recipes/terraform/install.go
+++ b/pkg/recipes/terraform/install.go
@@ -89,11 +89,10 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (*
 
 	// Verify Terraform installation is complete before proceeding
 	for attempt := 0; attempt <= installRetryCount; attempt++ {
-		errMsg := "Failed to verify Terraform installation: %s. Retrying after %d seconds"
 		_, _, err = tf.Version(ctx, false)
 		if err != nil {
 			if attempt < installRetryCount {
-				logger.Info(fmt.Sprintf(errMsg, err.Error(), retryDelaySecs))
+				logger.Info(fmt.Sprintf("Failed to verify Terraform installation: %s. Retrying after %d seconds", err.Error(), retryDelaySecs))
 				metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallVerificationDuration(ctx, installStartTime,
 					[]attribute.KeyValue{
 						metrics.TerraformVersionAttrKey.String("latest"),
@@ -103,7 +102,7 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (*
 				time.Sleep(time.Duration(retryDelaySecs) * time.Second)
 				continue
 			}
-			return nil, fmt.Errorf(errMsg, err.Error(), retryDelaySecs)
+			return nil, err
 		}
 	}
 

--- a/pkg/recipes/terraform/install.go
+++ b/pkg/recipes/terraform/install.go
@@ -34,9 +34,9 @@ import (
 )
 
 const (
-	installSubDir     = "install"
-	installRetryCount = 5
-	retryDelaySecs    = 3
+	installSubDir                     = "install"
+	installVerificationRetryCount     = 5
+	installVerificationRetryDelaySecs = 3
 )
 
 // Install installs Terraform under /install in the provided Terraform root directory for the resource. It installs
@@ -88,33 +88,32 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (*
 	}
 
 	// Verify Terraform installation is complete before proceeding
-	for attempt := 0; attempt <= installRetryCount; attempt++ {
+	for attempt := 0; attempt <= installVerificationRetryCount; attempt++ {
 		_, _, err = tf.Version(ctx, false)
 		if err == nil {
+			metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallVerificationDuration(ctx, installStartTime,
+				[]attribute.KeyValue{
+					metrics.TerraformVersionAttrKey.String("latest"),
+					metrics.OperationStateAttrKey.String(metrics.SuccessfulOperationState),
+				},
+			)
 			break
 		}
 		if err != nil {
-			if attempt < installRetryCount {
-				logger.Info(fmt.Sprintf("Failed to verify Terraform installation: %s. Retrying after %d seconds", err.Error(), retryDelaySecs))
+			if attempt < installVerificationRetryCount {
+				logger.Info(fmt.Sprintf("Failed to verify Terraform installation completion: %s. Retrying after %d seconds", err.Error(), installVerificationRetryDelaySecs))
 				metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallVerificationDuration(ctx, installStartTime,
 					[]attribute.KeyValue{
 						metrics.TerraformVersionAttrKey.String("latest"),
 						metrics.OperationStateAttrKey.String(metrics.FailedOperationState),
 					},
 				)
-				time.Sleep(time.Duration(retryDelaySecs) * time.Second)
+				time.Sleep(time.Duration(installVerificationRetryDelaySecs) * time.Second)
 				continue
 			}
-			return nil, err
+			return nil, fmt.Errorf("failed to verify Terraform installation completion after %d attempts. Error: %s", installVerificationRetryCount, err.Error())
 		}
 	}
-
-	metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallVerificationDuration(ctx, installStartTime,
-		[]attribute.KeyValue{
-			metrics.TerraformVersionAttrKey.String("latest"),
-			metrics.OperationStateAttrKey.String(metrics.SuccessfulOperationState),
-		},
-	)
 
 	// Configure Terraform logs once Terraform installation is complete
 	configureTerraformLogs(ctx, tf)

--- a/pkg/recipes/terraform/install.go
+++ b/pkg/recipes/terraform/install.go
@@ -90,6 +90,9 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (*
 	// Verify Terraform installation is complete before proceeding
 	for attempt := 0; attempt <= installRetryCount; attempt++ {
 		_, _, err = tf.Version(ctx, false)
+		if err == nil {
+			break
+		}
 		if err != nil {
 			if attempt < installRetryCount {
 				logger.Info(fmt.Sprintf("Failed to verify Terraform installation: %s. Retrying after %d seconds", err.Error(), retryDelaySecs))

--- a/pkg/recipes/terraform/install.go
+++ b/pkg/recipes/terraform/install.go
@@ -27,25 +27,28 @@ import (
 	"github.com/hashicorp/hc-install/product"
 	"github.com/hashicorp/hc-install/releases"
 	"github.com/hashicorp/hc-install/src"
+	"github.com/hashicorp/terraform-exec/tfexec"
 	"github.com/radius-project/radius/pkg/metrics"
 	"github.com/radius-project/radius/pkg/ucp/ucplog"
 	"go.opentelemetry.io/otel/attribute"
 )
 
 const (
-	installSubDir = "install"
+	installSubDir     = "install"
+	installRetryCount = 5
+	retryDelaySecs    = 3
 )
 
 // Install installs Terraform under /install in the provided Terraform root directory for the resource. It installs
 // the latest version of Terraform and returns the path to the installed Terraform executable. It returns an error
 // if the directory creation or Terraform installation fails.
-func Install(ctx context.Context, installer *install.Installer, tfDir string) (string, error) {
+func Install(ctx context.Context, installer *install.Installer, tfDir string) (*tfexec.Terraform, error) {
 	logger := ucplog.FromContextOrDiscard(ctx)
 
 	// Create Terraform installation directory
 	installDir := filepath.Join(tfDir, installSubDir)
 	if err := os.MkdirAll(installDir, 0755); err != nil {
-		return "", fmt.Errorf("failed to create directory for terraform installation for resource: %w", err)
+		return nil, fmt.Errorf("failed to create directory for terraform installation for resource: %w", err)
 	}
 
 	logger.Info(fmt.Sprintf("Installing Terraform in the directory: %q", installDir))
@@ -66,7 +69,7 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (s
 				metrics.OperationStateAttrKey.String(metrics.FailedOperationState),
 			},
 		)
-		return "", err
+		return nil, err
 	}
 
 	metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallationDuration(ctx, installStartTime,
@@ -78,5 +81,27 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (s
 
 	logger.Info(fmt.Sprintf("Terraform latest version installed to: %q", execPath))
 
-	return execPath, nil
+	// Create a new instance of tfexec.Terraform with current Terraform installation path
+	tf, err := NewTerraform(ctx, tfDir, execPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify Terraform installation is complete before proceeding
+	for attempt := 0; attempt <= installRetryCount; attempt++ {
+		errMsg := "Failed to verify Terraform installation: %s. Retrying after %d seconds"
+		_, _, err = tf.Version(ctx, false)
+		if err != nil {
+			if attempt < installRetryCount {
+				logger.Info(fmt.Sprintf(errMsg, err.Error(), retryDelaySecs))
+				time.Sleep(time.Duration(retryDelaySecs) * time.Second)
+				continue
+			}
+			return nil, fmt.Errorf(errMsg, err.Error(), retryDelaySecs)
+		}
+	}
+
+	configureTerraformLogs(ctx, tf)
+
+	return tf, nil
 }

--- a/pkg/recipes/terraform/install.go
+++ b/pkg/recipes/terraform/install.go
@@ -99,20 +99,18 @@ func Install(ctx context.Context, installer *install.Installer, tfDir string) (*
 			)
 			break
 		}
-		if err != nil {
-			if attempt < installVerificationRetryCount {
-				logger.Info(fmt.Sprintf("Failed to verify Terraform installation completion: %s. Retrying after %d seconds", err.Error(), installVerificationRetryDelaySecs))
-				metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallVerificationDuration(ctx, installStartTime,
-					[]attribute.KeyValue{
-						metrics.TerraformVersionAttrKey.String("latest"),
-						metrics.OperationStateAttrKey.String(metrics.FailedOperationState),
-					},
-				)
-				time.Sleep(time.Duration(installVerificationRetryDelaySecs) * time.Second)
-				continue
-			}
-			return nil, fmt.Errorf("failed to verify Terraform installation completion after %d attempts. Error: %s", installVerificationRetryCount, err.Error())
+		if attempt < installVerificationRetryCount {
+			logger.Info(fmt.Sprintf("Failed to verify Terraform installation completion: %s. Retrying after %d seconds", err.Error(), installVerificationRetryDelaySecs))
+			metrics.DefaultRecipeEngineMetrics.RecordTerraformInstallVerificationDuration(ctx, installStartTime,
+				[]attribute.KeyValue{
+					metrics.TerraformVersionAttrKey.String("latest"),
+					metrics.OperationStateAttrKey.String(metrics.FailedOperationState),
+				},
+			)
+			time.Sleep(time.Duration(installVerificationRetryDelaySecs) * time.Second)
+			continue
 		}
+		return nil, fmt.Errorf("failed to verify Terraform installation completion after %d attempts. Error: %s", installVerificationRetryCount, err.Error())
 	}
 
 	// Configure Terraform logs once Terraform installation is complete

--- a/pkg/recipes/terraform/types.go
+++ b/pkg/recipes/terraform/types.go
@@ -74,8 +74,6 @@ func NewTerraform(ctx context.Context, tfRootDir, execPath string) (*tfexec.Terr
 		return nil, fmt.Errorf("failed to initialize Terraform: %w", err)
 	}
 
-	configureTerraformLogs(ctx, tf)
-
 	return tf, nil
 }
 

